File: cpp/src/ch05/alpha_beta.cc
#include "alpha_beta.h"
#include "time_keeper.h"

ScoreType alpha_beta_score(const State &state,
                           ScoreType alpha,
                           const ScoreType beta,
                           const int depth,
                           const TimeKeeper &time_keeper)
{
    // if time_over, we won't use the score anyway
    if (time_keeper.is_time_over())
        return alpha;

    if (state.is_done() || depth == 0)
        return state.get_score();

    auto legal_actions = state.legal_actions();
    if (legal_actions.empty())
        return state.get_score();

    for (const auto action : legal_actions)
    {
        State next_state = state;
        next_state.advance(action);

        ScoreType score = -alpha_beta_score(next_state, -beta, -alpha, depth - 1, time_keeper);

        if (score > alpha)
            alpha = score;
        if (alpha >= beta)
            return alpha;
    }
    return alpha;
}

int alpha_beta_action(const State &state,
                      const int depth,
                      const TimeKeeper &time_keeper)
{
    ScoreType best_action = 0; // set default as 0 for time_over case
    ScoreType alpha = -INF;
    ScoreType beta = INF;
    for (const auto action : state.legal_actions())
    {
        State next_state = state;
        next_state.advance(action);
        ScoreType score = -alpha_beta_score(next_state, -beta, -alpha, depth, time_keeper);
        if (score > alpha)
        {
            best_action = action;
            alpha = score;
        }
        if (time_keeper.is_time_over())
            break;
    }
    return best_action;
}

File: rust/src/ch05/alpha_beta.rs
use crate::base::alternate::{ActionFunc, AlternateState, Evaluatable};
use std::sync::Arc;

fn alpha_beta_score<T: AlternateState + Evaluatable>(
    initial_state: &T,
    mut alpha: isize,
    beta: isize,
    depth: usize,
) -> isize {
    if initial_state.is_done() || depth == 0 {
        return initial_state.evaluation();
    }

    let legal_actions = initial_state.legal_actions();
    if legal_actions.is_empty() {
        return initial_state.evaluation();
    }

    for action in legal_actions {
        let mut state = initial_state.clone();
        state.advance(action);
        let score = -alpha_beta_score(&state, -beta, -alpha, depth - 1);
        if score > alpha {
            alpha = score;
        }
        if alpha >= beta {
            return alpha;
        }
    }

    alpha
}

// utility to track score and action
#[derive(Debug, Clone)]
struct ScoreAction {
    score: isize,
    action: usize,
}

fn alpha_beta_action<T: AlternateState + Evaluatable>(
    initial_state: &T,
    depth: usize,
) -> usize {
    let mut score_actions = Vec::new();
    // to prevent overflow, make 1 smaller than limit
    let alpha = isize::MIN + 1;
    let beta = isize::MAX - 1;
    let legal_actions = initial_state.legal_actions();
    for action in legal_actions {
        let mut state = initial_state.clone();
        state.advance(action);
        let score = -alpha_beta_score(&state, -beta, -alpha, depth);
        score_actions.push(ScoreAction { score, action });
    }

    let best = score_actions.iter().max_by_key(|p| p.score).unwrap();
    best.action
}

pub fn alpha_beta_arc<T: AlternateState + Evaluatable>(
    depth: usize,
) -> ActionFunc<T> {
    Arc::new(move |state| -> usize { alpha_beta_action(state, depth) })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::base::alternate::MazeParams;
    use crate::ch05::maze_state;

    fn setup() -> maze_state::AlternateMazeState {
        let params = MazeParams {
            height: 3,
            width: 3,
            end_turn: 4,
        };
        maze_state::AlternateMazeState::new(0, params)
    }

    #[test]
    fn test_alpha_beta_arc() {
        let state = setup();
        let actual = alpha_beta_arc(4)(&state);
        let expected = 0;
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_alpha_beta_action() {
        let state = setup();
        let actual = alpha_beta_action(&state, 4);
        let expected = 0;
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_alpha_beta_score() {
        let state = setup();
        let alpha = isize::MIN + 1;
        let beta = isize::MAX - 1;
        let actual = alpha_beta_score(&state, alpha, beta, 3);
        let expected = 3;
        assert_eq!(actual, expected);
    }
}
